presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java

```diff
@@ -325,133 +418,291 @@ private void calculateVariableValuesForLongConstraints(BoundVariables.Builder va
+            case VARIABLE: {
+                String variableName = parameter.getVariable();
+                checkState(boundVariables.containsLongVariable(variableName),
+                        "Variable is not bound: %s", variableName);
+                Long variableValue = boundVariables.getLongVariable(variableName);
+                return TypeSignatureParameter.of(variableValue);
+            }
+            case LONG: {
+                return parameter;
             }
+            default:
+                throw new IllegalStateException("Unknown parameter kind: " + parameter.getKind());
         }
-        return true;
     }
 
-    private boolean allTypeVariablesBound(BoundVariables boundVariables)
+    private static List<TypeSignature> expandVarargFormalTypeSignature(List<TypeSignature> formalTypeSignatures, int actualArity)
     {
-        return boundVariables.getTypeVariables().keySet().equals(typeVariableConstraints.keySet());
+        int variableArityArgumentsCount = actualArity - formalTypeSignatures.size() + 1;
+        if (variableArityArgumentsCount == 0) {
+            return formalTypeSignatures.subList(0, formalTypeSignatures.size() - 1);
+        }
+        if (variableArityArgumentsCount == 1) {
+            return formalTypeSignatures;
+        }
+        checkArgument(variableArityArgumentsCount > 1 && !formalTypeSignatures.isEmpty());
+
+        ImmutableList.Builder<TypeSignature> builder = ImmutableList.builder();
+        builder.addAll(formalTypeSignatures);
+        TypeSignature lastTypeSignature = formalTypeSignatures.get(formalTypeSignatures.size() - 1);
+        for (int i = 1; i < variableArityArgumentsCount; i++) {
+            builder.add(lastTypeSignature);
+        }
+        return builder.build();
     }
 
-    public static Signature applyBoundVariables(Signature signature, BoundVariables boundVariables, int arity)
+    private interface TypeConstraintSolver
     {
-        List<TypeSignature> argumentSignatures = fillInMissingVariableArguments(signature.getArgumentTypes(), arity);
-        List<TypeSignature> boundArgumentSignatures = applyBoundVariables(argumentSignatures, boundVariables);
-        TypeSignature boundReturnTypeSignature = applyBoundVariables(signature.getReturnType(), boundVariables);
+        SolverReturnStatus update(BoundVariables.Builder bindings);
+    }
 
-        return new Signature(
-                signature.getName(),
-                signature.getKind(),
-                ImmutableList.of(),
-                ImmutableList.of(),
-                boundReturnTypeSignature,
-                boundArgumentSignatures,
-                false
-        );
+    private enum SolverReturnStatus
+    {
+        UNCHANGED_SATISFIED,
+        UNCHANGED_NOT_SATISFIED,
+        CHANGED,
+        UNSOLVABLE,
     }
 
-    private static List<TypeSignature> fillInMissingVariableArguments(List<TypeSignature> argumentSignatures, int arity)
+    private class SolverReturnStatusMerger
```

Comment:
why do you need a separate class for this logic? Wouldn't simple method `mergeSolverReturnStatuses` be enough?